[WebAssembly] Enable nontrapping-fptoint and bulk-memory by default.

sunfishcode opened the pull request:
We were prepared to enable these features [back in February], but they got pulled for what appear to be unrelated reasons. So let's have another try at enabling them!

Another motivation here is that it'd be convenient for the [Trail1 proposal] if "trail1" is a ~~superset~~subset of "generic". Edit: except for extended-const though.

[back in February]: https://github.com/WebAssembly/tool-conventions/issues/158#issuecomment-1931119512
[Trail1 proposal]: https://github.com/llvm/llvm-project/pull/112035

sunfishcode committed:
[WebAssembly] Enable nontrapping-fptoint and bulk-memory by default.

We were prepared to enable these features [back in February], but they
got pulled for what appear to be unrelated reasons. So let's have another
try at enabling them!

Another motivation here is that it'd be convenient for the
[Trail1 proposal] if "trail1" is a superset of "generic".

[back in February]: https://github.com/WebAssembly/tool-conventions/issues/158#issuecomment-1931119512
[Trail1 proposal]: https://github.com/llvm/llvm-project/pull/112035

sunfishcode committed:
Update tests.

sunfishcode committed:
Update tests.

sunfishcode committed:
Update tests.

sbc100 commented on lld/test/wasm/lto/libcall-archive.ll:
```diff
@@ -2,7 +2,7 @@
 ; RUN: llvm-as -o %t2.o %S/Inputs/libcall-archive.ll
 ; RUN: llvm-ar rcs %t.a %t2.o
-; RUN: wasm-ld -o %t %t.o %t.a
+; RUN: wasm-ld -mllvm -mattr=-bulk-memory -o %t %t.o %t.a
```
Why is this needed?  Should this come from the object file if it needs it?

sunfishcode replied:
This is an LTO test, so there is no `clang` or `llvm-mc` to pass `-mattr=` to. I guess we could put `-bulk-memory` in function attributes in the .ll files, if you'd prefer that.

sbc100 replied:
I was thinking that might make more sense yes.   Seems like that would match better what llvm lto would produce.  I don't feel strongly though.

sunfishcode replied:
Done

sunfishcode committed:
Move `-bulk-memory` flags from wasm-ld flags to a function attributes.

aheejin commented on clang/test/Preprocessor/wasm-target-features.c:
```diff
@@ -166,6 +166,8 @@
 // GENERIC-INCLUDE-DAG: #define __wasm_mutable_globals__ 1{{$}}
 // GENERIC-INCLUDE-DAG: #define __wasm_reference_types__ 1{{$}}
 // GENERIC-INCLUDE-DAG: #define __wasm_sign_ext__ 1{{$}}
+// GENERIC-INCLUDE-DAG: #define __wasm_nontrapping_fptoint__ 1{{$}}
+// GENERIC-INCLUDE-DAG: #define __wasm_bulk_memory__ 1{{$}}
```
Nit: Probably my OCD, but I've trying to maintain the sorted order of the features... Can we move these lines into the sorted order?

sunfishcode replied:
Done

aheejin commented on llvm/test/CodeGen/WebAssembly/cfg-stackify-eh-legacy.ll:
```diff
@@ -1,9 +1,9 @@
-; RUN: llc < %s -O0 -disable-wasm-fallthrough-return-opt -wasm-disable-explicit-locals -wasm-keep-registers -verify-machineinstrs -wasm-enable-eh -exception-model=wasm -mattr=+exception-handling | FileCheck %s --check-prefix=NOOPT
-; RUN: llc < %s -disable-wasm-fallthrough-return-opt -wasm-disable-explicit-locals -wasm-keep-registers -disable-block-placement -verify-machineinstrs -fast-isel=false -machine-sink-split-probability-threshold=0 -cgp-freq-ratio-to-skip-merge=1000 -wasm-enable-eh -exception-model=wasm -mattr=+exception-handling -wasm-disable-ehpad-sort -stats 2>&1 | FileCheck %s --check-prefix=NOSORT
-; RUN: llc < %s -disable-wasm-fallthrough-return-opt -disable-block-placement -verify-machineinstrs -fast-isel=false -machine-sink-split-probability-threshold=0 -cgp-freq-ratio-to-skip-merge=1000 -wasm-enable-eh -exception-model=wasm -mattr=+exception-handling -wasm-disable-ehpad-sort | FileCheck %s --check-prefix=NOSORT-LOCALS
+; RUN: llc < %s -disable-wasm-fallthrough-return-opt -wasm-disable-explicit-locals -wasm-keep-registers -disable-block-placement -verify-machineinstrs -fast-isel=false -machine-sink-split-probability-threshold=0 -cgp-freq-ratio-to-skip-merge=1000 -wasm-enable-eh -exception-model=wasm -mattr=+exception-handling,bulk-memory | FileCheck %s
```
Why do we need to enable bulk-memory here?

sunfishcode replied:
The `mem_intrinsics` test calls `memcpy`/etc. and expects them to be calls rather than expanded inline to `memory.copy`/etc.

aheejin commented on lld/test/wasm/custom-section-name.ll:
```diff
@@ -1,4 +1,4 @@
-; RUN: llc -filetype=obj %s -o %t.o
+; RUN: llc -filetype=obj -mattr=-bulk-memory %s -o %t.o
```
Why do we need to disable bulk memory here and in many other files?

sunfishcode replied:
In lld/test/wasm/custom-section-name.ll, it's because bulk-memory makes the target_features section bigger and thus changes some of the section offsets and it was easier to just disable bulk-memory than to figure out the new section offsets. I can update the section offsets if you prefer though.

In lld/test/wasm/data-segments.ll, it's because the test is written to test both bulk-memory and non-bulk-memory, and to keep it doing that, it's now necessary to explicitly disable bulk-memory for the non-bulk-memory case.

In lld/test/wasm/lto/stub-library-libcall.s it's because the test is testing the use of a memcpy stub function, and with bulk-memory it gets inlined.

llvm/test/MC/WebAssembly/extern-functype-intrinsic.ll is similarly testing memset calls.

llvm/test/MC/WebAssembly/libcall.ll is similarly testing memcpy calls.

aheejin replied:
Thanks for the explanation!

sunfishcode replied:
I've now investigated custom-section-name.ll more.

With bulk-memory, we get a `__wasm_init_memory` function:
```
(module $custom-section-name.bss.wasm
  (type (;0;) (func))
  (import "env" "memory" (memory (;0;) 2))
  (func $__wasm_call_ctors (;0;) (type 0))
  (func $__wasm_init_memory (;1;) (type 0)
    i32.const 1036
    i32.const 0
    i32.const 4
    memory.fill
  )
  (global $__stack_pointer (;0;) (mut i32) i32.const 66576)
  (start $__wasm_init_memory)
  (data $WowZero! (;0;) (i32.const 1024) "\00\00\00\00")
  (data $MyAwesomeSection (;1;) (i32.const 1028) "*\00\00\00")
  (data $AnotherGreatSection (;2;) (i32.const 1032) "\07\00\00\00")
  (@custom "target_features" (after data) "\06+\0bbulk-memory+\16call-indirect-overlong+\0amultivalue+\0fmutable-globals+\0freference-types+\08sign-ext")
)
```
Without bulk-memory we get a .bss section:
```
(module $custom-section-name.bss.wasm
  (type (;0;) (func))
  (import "env" "memory" (memory (;0;) 2))
  (func $__wasm_call_ctors (;0;) (type 0))
  (global $__stack_pointer (;0;) (mut i32) i32.const 66576)
  (data $WowZero! (;0;) (i32.const 1024) "\00\00\00\00")
  (data $MyAwesomeSection (;1;) (i32.const 1028) "*\00\00\00")
  (data $AnotherGreatSection (;2;) (i32.const 1032) "\07\00\00\00")
  (data $.bss (;3;) (i32.const 1036) "\00\00\00\00")
  (@custom "target_features" (after data) "\05+\16call-indirect-overlong+\0amultivalue+\0fmutable-globals+\0freference-types+\08sign-ext")
)
```

The test is intended to test the use of the .bss section, and we only get a .bss section in this test if we disable bulk-memory.

dschuff replied:
Hm, interesting, so we are using passive segments anytime bulk memory is enabled. That makes sense since this was all meant for threads, but is this actually an improvement in the non-threads case? i.e. maybe we don't want the extra explicit initialization code if we can help it?

edit: i just checked this out but I'll leave it here FTR: it looks like after linking we actually don't use passive segments unless threads is also enabled. so this seems like not a problem.

sunfishcode committed:
Delete redundant features.

sunfishcode committed:
Sort the feature macros alphabetically.

sunfishcode commented on clang/lib/Basic/Targets/WebAssembly.cpp:
```diff
@@ -151,8 +151,10 @@ bool WebAssemblyTargetInfo::initFeatureMap(
     llvm::StringMap<bool> &Features, DiagnosticsEngine &Diags, StringRef CPU,
     const std::vector<std::string> &FeaturesVec) const {
   auto addGenericFeatures = [&]() {
+    Features["bulk-memory"] = true;
```
Done

aheejin reviewed (approved):
This LGTM, but I guess we need to wait until Emscripten disables them by default, because there are some failing Emscripten tests and Binaryen doesn't yet have lowering passes for these. I don't think we need to block this on all these; we would just need to disable these two feature in Emscripten by default in the meantime.

sunfishcode commented:
I've now added a patch which fixes https://github.com/llvm/llvm-project/issues/63755.

aheejin commented:
> I've now added a patch which fixes #63755.

Can you split this patch into a separate PR?

sunfishcode commented:
> Can you split this patch into a separate PR?

Sure, I've now created https://github.com/llvm/llvm-project/pull/112617 with just the that patch.

sunfishcode committed:
Fix subtarget feature coalescing to allow default features to be disabled.

Fix disabling "bulk-memory" via function attributes.

 - Fix FixFunctionBitcasts to add attributes to the generated wrapper
   functions so that they don't pick up default target features.

 - Fix `coalesceFeatures` to only consider defined functions.

 - Fix `coalesceFeatures` to combine the union starting from the empty
   set, rather than starting from the target machine set, so that if
   some feature is explicitly disabled in all functions, it's disabled
   in the result.

This fixes the wasm/lto/libcall-archive.ll test once bulk-memory is
enabled, as that test depends on being able to disable bulk-memory
via function attributes.

sunfishcode committed:
code formatting

sunfishcode committed:
formatting

dschuff reviewed (approved):
I think Emscripten is ready for this now, thanks for working on this!

sunfishcode commented:
Thanks!
